spinnaker-monitoring-daemon/spinnaker-monitoring/newrelic_service.py

```diff
@@ -0,0 +1,116 @@
+        self.tags = tags
+
+    def parse_metric(self, service, metric_name, metric_instance, metric_data, service_data, metric_list):
+        kind = self.spectator_helper.determine_primitive_kind(
+            metric_data["kind"])
+        for metric_data_value in metric_instance["values"]:
+            tags = self.tags.copy()
+            for tag in metric_instance["tags"]:
+                tags[tag["key"]] = tag["value"]
+            tags["applicationName"] = service_data["applicationName"]
+            tags["applicationVersion"] = service_data["applicationVersion"]
+            interval = metric_data_value["t"] - \
+                service_data["__collectStartTime"]
+            if kind == spectator_client.GAUGE_PRIMITIVE_KIND:
+                metric_list.append(GaugeMetric(
+                    name=metric_name, value=metric_data_value["v"], tags=tags))
+            else:
+                metric_list.append(CountMetric(
+                    name=metric_name, value=metric_data_value["v"], interval_ms=interval, tags=tags))
+
+    def publish_metrics(self, service_metrics):
+        metric_list = []
+        spectator_client.foreach_metric_in_service_map(
+            service_metrics, self.parse_metric, metric_list)
+        # using 1000 as a known-good size, not a theoretical maximum
+        chunk_size = 1000
+        chunks = [metric_list[i:i + chunk_size]
+                  for i in xrange(0, len(metric_list), chunk_size)]
+        for chunk in chunks:
+            response = self.metric_client.send_batch(chunk)
+            response.raise_for_status()
+        return len(metric_list)
+
+
+NEWRELIC_KUBERNETES_METADATA_MAPPING = {
+    "NEW_RELIC_METADATA_KUBERNETES_CLUSTER_NAME": "clusterName",
+    "NEW_RELIC_METADATA_KUBERNETES_NODE_NAME": "nodeName",
+    "NEW_RELIC_METADATA_KUBERNETES_NAMESPACE_NAME": "namespaceName",
+    "NEW_RELIC_METADATA_KUBERNETES_DEPLOYMENT_NAME": "deploymentName",
+    "NEW_RELIC_METADATA_KUBERNETES_POD_NAME": "podName",
+    "NEW_RELIC_METADATA_KUBERNETES_CONTAINER_NAME": "containerName",
+    "NEW_RELIC_METADATA_KUBERNETES_CONTAINER_IMAGE_NAME": "containerImageName",
+}
+
+
+def extract_tags(options):
+    tags = {}
+    if 'newrelic' in options and 'tags' in options['newrelic']:
+        option_tags = options['newrelic'].get('tags', []) or []
+        for tag in option_tags:
+            if tag.find(":"):
+                tags.update(tag.split(":", 2))
+    for env_var_name, tag_name in NEWRELIC_KUBERNETES_METADATA_MAPPING.iteritems():
+        if env_var_name in os.environ:
+            tags[tag_name] = os.environ[env_var_name]
+
+    return tags
+
+
+def make_new_relic_service(options, spectator_helper=None):
+    spectator_helper = spectator_helper or spectator_client.SpectatorClientHelper(
+        options)
+    if 'newrelic' in options and 'insert_key' in options['newrelic']:
+        insert_key = options['newrelic']['insert_key']
+    elif 'NEWRELIC_INSERT_KEY' in os.environ:
```

Comment:
Dont environment variables usually have higher precedence than config options?
Or are these options command-line ?

Reply:
Yeah, the config file is overwriting environment variables here. Should probably be the opposite.